core/state/statedb.go

```diff
@@ -853,17 +865,42 @@ func (db *DB) UpdateValidatorWrapper(
 	if err := val.SanityCheck(); err != nil {
 		return err
 	}
-
 	by, err := rlp.EncodeToBytes(val)
 	if err != nil {
 		return err
 	}
+	// has revert in-built for the code field
 	db.SetCode(addr, by)
 	// update cache
 	db.stateValidators[addr] = val
 	return nil
 }
 
+// UpdateValidatorWrapperWithRevert updates staking information of
+// a given validator (including delegation info)
+func (db *DB) UpdateValidatorWrapperWithRevert(
+	addr common.Address, val *stk.ValidatorWrapper,
+) error {
+	if err := val.SanityCheck(); err != nil {
+		return err
+	}
+	// a copy of the existing store can be used for revert
+	// since we are replacing the existing with the new anyway
+	prev, err := db.ValidatorWrapper(addr, true, false)
```

Comment:
Should this be "db.ValidatorWrapper(addr, false, true)"? Since you want a copy to be stored as the change journal?

Reply:
This is because the caller is sending a copy, which is being added to the `db`, while the original is in the `journal`.

Reply:
I see. make sense, the original is replaced with the new one, so it's safe to directly use it without copying. thanks